JS Learning Area updates: looping code

wbamberg opened the pull request:
Part of https://github.com/mdn/content/issues/10337.

This PR updates the "Looping code" part of the JS Learning Area: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Looping_code.

The main thing here is to:
* introduce `for...of` as the recommended way to iterate through a collection
* briefly introduce `map()` and `filter()`
* make sure all worked examples which can/should used `for...of` are doing so.

It's a bit tricky because we have already introduced `for...of`, `map()`, and `filter()` in the [Arrays](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/Arrays) page, but I thought it's going to be easiest for readers just to present it again here.

I've also made a tiny update to the https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Test_your_skills:_Loops page, to reflect updates to the learning area in https://github.com/mdn/learning-area/pull/421.

wbamberg committed:
JS Learning Area updates: looping code

wbamberg committed:
Update Test your skills page on loops

hamishwillee commented on files/en-us/learn/javascript/building_blocks/looping_code/index.md:
```diff
@@ -297,15 +428,17 @@ btn.addEventListener('click', function() {
   let searchName = input.value.toLowerCase();
```
const ?

hamishwillee commented:
Is @teoli2003 on holiday? I'm hesitant to review if you can get someone with more expertise.

wbamberg committed:
let->const

wbamberg commented:
> Is @teoli2003 on holiday? I'm hesitant to review if you can get someone with more expertise.

It's up to you, I'm happy for you to review this if you want, I'm sure you have plenty of expertise. Also this is introductory so reviewing for clarity and consistency are particularly important, and there shouldn't be anything technically very complex here.

hamishwillee commented on files/en-us/learn/javascript/building_blocks/looping_code/index.md:
```diff
@@ -174,9 +174,91 @@ ctx.fill();
+2. Assign it to the variable `cat` and then run the code between the curly brackets `{}`.
+3. Get the next item, and repeat (2) until you've reached the end of the collection.
+
+Note that we have to use `let` for the `cat` variable, because we're assigning it to a new item each time we go round the loop.
```
Is this true? My understanding was that you could use const, let, or var because the scope of cat (for let and const) is within the loop. You can't use const if you then want to modify cat in the loop.

hamishwillee replied:
If so it _might_ be worth modifying item 2 above. This kind of implies the variable is created once and reassigned to a new value on each loop. I think that is true for var, but not for let.

wbamberg replied:
You're right! Apparently I don't understand const either!

hamishwillee committed:
Mostly just very minor layout changes

hamishwillee commented:
@wbamberg Really good. I am happy to merge as is except for my comment above https://github.com/mdn/content/pull/10645#discussion_r757160664

In terms of other things I noticed that we might want to look at
- up the top "keeping me in the loop" example is all about the normal for loop and we don't mention `for...of`. It all makes sense, but you might want to consider introducing the concept earlier.
- The discussion of while/do and do/while is all about the syntax and they are kind of presented as equal alternatives. My understanding is that the difference is that in one syntax the loop is always executed at least once while in the other it is only executed on the condition. That's a significant thing to note because it affects when you would choose them. Perhaps I'm thinking of another programming language :-).

Last of all, I did inject some layout changes which you can ignore. I just do this because shorter lines are easier to review in github diffs. I tend to break on sentences.

hamishwillee commented on files/en-us/learn/javascript/building_blocks/looping_code/index.md:
```diff
@@ -40,7 +40,9 @@ Programming languages are very useful for rapidly completing repetitive tasks, f
-Loops, loops, loops. As well as being associated with [popular breakfast cereals](https://en.wikipedia.org/wiki/Froot_Loops), [roller coasters](https://en.wikipedia.org/wiki/Vertical_loop), and [musical production](<https://en.wikipedia.org/wiki/Loop_(music)>), they are also a critical concept in programming. Programming loops are all to do with doing the same thing over and over again, which is  ...
+Loops, loops, loops.
+As well as being associated with [popular breakfast cereals](https://en.wikipedia.org/wiki/Froot_Loops), [roller coasters](https://en.wikipedia.org/wiki/Vertical_loop), and [musical production](<https://en.wikipedia.org/wiki/Loop_(music)>), they are also a critical concept in programming.
+Programming loops are all to do with doing the same thing over and over again, which is called **iteration** in "programming speak".
```
FYI this is an actual minor change.

wbamberg commented:
> @wbamberg Really good. I am happy to merge as is except for my comment above [#10645 (comment)](https://github.com/mdn/content/pull/10645#discussion_r757160664)
> 
> In terms of other things I noticed that we might want to look at
> 
>     * up the top "keeping me in the loop" example is all about the normal for loop and we don't mention `for...of`. It all makes sense, but you might want to consider introducing the concept earlier.

Yeah, I wasn't sure what to do about that. I couldn't really see how to adapt it to make it more `for...of` friendly. I rationalized keeping it as is by thinking that `for..of` loops have the same things (initialization, increment, condition), it's just they're not exposed. But maybe it would be better without all that farmer stuff?

> 
>     * The discussion of while/do and do/while is all about the syntax and they are kind of presented as equal alternatives. My understanding is that the difference is that in one syntax the loop is always executed at least once while in the other it is only executed on the condition. That's a significant thing to note because it affects when you would choose them. Perhaps I'm thinking of another programming language :-).
> 

Yes, good point. I didn't look too closely at those, but while I am here it's worth adding that.

hamishwillee commented:
> But maybe it would be better without all that farmer stuff?

I would. I'm not sure it adds anything over what is in the "why bother" section. Also someone is going to complain about the farmer being a "he" one day.

wbamberg committed:
Use const for for...of loop variable

wbamberg committed:
Remove farmer

wbamberg committed:
Clarify difference between while and do...while loops

wbamberg commented:
OK, I have:
* fixed `for...of` loops to use `const`: https://github.com/mdn/content/pull/10645/commits/ddc7851138407b92a77b130172e207b25f1e61ea
* removed the farmer stuff (also opted to show all the JS for the inital "drawing circles" example, I thought it was helpful): https://github.com/mdn/content/pull/10645/commits/bd1a60d96724b1544750b5d3757c30a8658ac4b6
* clarified that a `do...while` loop always runs at least once (it did actually say this, but I rewrote it to make is more obvious). I also updated the examples to be complete runnable snippets: https://github.com/mdn/content/pull/10645/commits/1348ae903b51a5fb54c359756ca61a27d3b2fc03

wbamberg committed:
Fix a typo

wbamberg committed:
let->const

hamishwillee reviewed (approved):
Yes, heaps better. Particularly the removal of the farmer stuff at the top - that just felt like wasted time.

I was going to make a comment that we should _also_ update "Which loop type should you use?" to mention that you use "do...while` loop if you have to run it once. But not sure that was needed - what is there now captures the flavour/intent.
